Publish symbols using release pipelines

JohnTortugo opened the pull request:
Closes: https://github.com/dotnet/arcade/issues/1786

Bulk of the changes here is copying the `PublishSymbols` target from Publish.proj into a separate file that will be used by a release pipeline to publish symbols to MSDL & SymWeb.

I did other changes around formatting / commenting.

tmat commented on eng/common/PublishToPackageFeed.proj:
```diff
@@ -1,11 +1,13 @@
-  informed target feeds.
--->
+<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
 <Project Sdk="Microsoft.NET.Sdk">
```
>Sdk="Microsoft.NET.Sdk" [](start = 9, length = 23)

Why are these SDK projects?

JohnTortugo replied:
Talked with @tmat. He said to leave it for now.

chcosta replied:
I haven't read most of the comments in this PR, and this has probably been discussed elsewhere; but this still feels like abusing the system to be distributing these projects via eng/common files.

JohnTortugo replied:
@tmat has made that observation before. A more immediate solution would be to move these files to somewhere under the SDK. I've an issue to track that: https://github.com/dotnet/arcade/issues/2371 I discussed a long term solution yesterday with @tmat and he shared some nice ideas. I need to put them on an issue, though. 

That being said I'd love to hear more about the direction this should take. I'll ping you later or if you prefer plz post them here.

jcagme commented on eng/common/PublishToSymbolServers.proj:
```diff
@@ -0,0 +1,75 @@
+
+    <PropertyGroup>
+      <DotNetSymbolExpirationInDays Condition="'$(DotNetSymbolExpirationInDays)' == ''">3650</DotNetSymbolExpirationInDays>
+      
```
Nit: remove empty line

mmitche commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:
```diff
@@ -195,9 +195,28 @@
+        https://github.com/dotnet/core-eng/issues/3645
+      -->
+    <ItemGroup>
+      <FilesToPublishToSymbolServer Include="$(ArtifactsSymStoreDirectory)**\*.pdb"/>
```
What about symbol packages? Are they handled?  IIRC we end up unpacking the .symbol.nupkgs produced by the build and publishing the PDBs.

JohnTortugo replied:
AFAIU both .pdbs and .symbols.nupkgs get published to symbol servers - handled by the `PublishSymbols` target/task. `.symbol.nupkgs` are also published to package feeds - handled by the same target/task that publishes regular packages.

jcagme commented:
I'd look at how corefx is doing it. I know they end up publishing to those servers so we'd just need to see what they do/how

dagood commented:
> Bulk of the changes here is copying the `PublishSymbols` target from Publish.proj into a separate file

Why? Why was it already copied into `Publish.proj` as well? How is this maintainable?

> I think most repositories are publishing symbol.nupkg files are then those are getting unpacked and pushed to symweb/msdl. Are those covered here?
> 
> @dagood Is this your understanding?

The `PublishSymbols` task does the digging around in the symbol package (zip) file.

The symbol packages do also need to be published to a build storage location so they can be reuploaded later at release time to extend the expiration time. (Not sure if that's what you were saying.)

dagood commented on eng/common/PublishToSymbolServers.proj:
```diff
@@ -0,0 +1,74 @@
+                    ExpirationInDays="$(DotNetSymbolExpirationInDays)"
+                    VerboseLogging="true"
+                    DryRun="false"
+                    ConvertPortablePdbsToWindowsPdbs="false"
```
This doesn't satisfy requirements. It might be ok to turn off conversion temporarily while someone works on the failure to unblock daily build flow, but it must be done for some debuggers to work.

JohnTortugo replied:
I don't know which requirement is that and I actually didn't disable this. It's just a copy of what Arcade currently does. I'm not aware of an issue to re-enable this conversion, but sounds like there should be one, right? /cc @tmat who seems to have been the last one to touch that part of the SDK.

dagood replied:
I edited it into https://github.com/dotnet/arcade/issues/167#issuecomment-478780348 shortly after posting, sorry if you saw an earlier version there:

> Windows PDB file is not included [in the symbol package itself]: it is converted on demand at upload time and should be checked by availability validation.

In general the availability rule comes from https://osgwiki.com/wiki/Symbol_Server_-_Symbols_Archiving_and_Indexing#Expectations_for_Each_Product_Group. Microsoft still depends on `symchk`, which as mentioned in https://github.com/dotnet/arcade/issues/1874#issuecomment-457361857 doesn't work on portable PDBs.

Yes, this should be turned on, it may impact pre-shipping manual validation testing if those tests are extensive enough.

JohnTortugo replied:
Thanks for clarifying that @dagood! Now I get it. I'll see to enable this on the current PR. It would be good to know why did that get disabled in the first place. Do you know if the code from `PublishSymbols` target in Publish.proj is a copy from somewhere else?

dagood replied:
Thanks. I don't doubt some repo failed with it turned on and you may need a configuration setting for that. About copying: I don't know the history of the target but I do know it currently duplicates the targets file in the package. (More info in main thread about this.)

JohnTortugo replied:
Just saw your other message. Thanks.

tmat replied:
@dagood Arcade SDK converts the PDBs during build, not when publishing them. We do satisfy all requirements on symbols - note that Roslyn inserts into VS and requires Watson to work as well.

tmat replied:
See https://github.com/dotnet/arcade/pull/2407/files/c2515fda3f813ac7b7d91ecf3008587c1ca68fd5#diff-a175e2256272e2c2793762edb8c302aaR205

dagood replied:
@tmat Interesting... how are the converted PDBs published to symbol servers and preserved to republish for releases? I'll edit https://github.com/dotnet/arcade/issues/167#issuecomment-478780348 with the new info.

dagood replied:
@tmat How are symbols produced this way stored for republish (to extend the expiration date)? Are we retaining the build for the artifact, e.g. a zip of the PDBs?

I'll update https://github.com/dotnet/arcade/issues/167#issuecomment-478780348 with the Arcade way.

tmat replied:
I don't think we are, or we really need to. If we want to extend expiration data we should be able to get the symbol information directly from the symbol server. 

Not sure if converting on publish is actually desirable. We publish to two symbol servers - it's wasteful to run the conversion twice (it's actually not very fast). Also, the conversion may fail. If it fails, then we end up with some symbols published and some not. It's better to separate conversion and publishing in two steps.

tmat replied:
That said there is definitely work to be done to support extending expiration dates. For now we have been publishing everything with 10 year expiration so that we don't need to extend.

dagood replied:
> I don't think we are, or we really need to. If we want to extend expiration data we should be able to get the symbol information directly from the symbol server.

Filed https://github.com/dotnet/arcade/issues/2415 to track this down. We've been doing this for .NET Core Runtime releases manually: we run a manual symbol publishing step on the symbol packages that has PDB conversion enabled.

> Not sure if converting on publish is actually desirable.

I don't think it is either, it was mainly just (highly) convenient when we didn't have a lot of power to put new things in the infra, pre-Arcade.

dagood replied:
> That said there is definitely work to be done to support extending expiration dates. For now we have been publishing everything with 10 year expiration so that we don't need to extend.

30 years is the minimum I heard about, maybe that's runtime rather than Visual Studio.

tmat replied:
I have never heard of 30 years. That seems excessive given that .NET Core support is 5 years or something like that.

tmat replied:
Can we actually first get the real requirements, not just what we heard? 
Also, I suggest we just publish all symbols with the max expiration time we need and don't do any extending. Much simpler and less error prone that way. Storage is cheap.

dagood replied:
I wish I knew where to find requirements other than https://github.com/dotnet/core-eng/issues/3382. I'm on board with finding that before taking action on any of this.

JohnTortugo replied:
> Not sure if converting on publish is actually desirable. We publish to two symbol servers - it's wasteful to run the conversion twice (it's actually not very fast). Also, the conversion may fail. If it fails, then we end up with some symbols published and some not. It's better to separate conversion and publishing in two steps.

Agree. I won't enable the conversion in publishing.

JohnTortugo commented:
> Why? Why was it already copied into `Publish.proj` as well? How is this maintainable?
> 
> >

@dagood the duplication should be only temporary. We are in the process of creating a drop-in replacement for the way Arcade publish packages. At some point of the work we'll remove the `PublishSymbols` target from Publish.proj. See plan here: https://github.com/dotnet/arcade/blob/master/Documentation/CorePackages/AsyncPublishing.md

dagood commented:
> At some point of the work we'll remove the `PublishSymbols` target from Publish.proj.

Will it be removed from `PublishToSymbolServers.proj` too to get rid of duplication from the symbol upload package's [PublishSymbols.targets](https://devdiv.visualstudio.com/DevDiv/_git/DotNet-SymUploader?path=%2Fsrc%2FMicrosoft.SymbolUploader.Build.Task%2FPackageFiles%2FPublishSymbols.targets&version=GBmaster)?

JohnTortugo commented:
> Will it be removed from `PublishToSymbolServers.proj` too to get rid of duplication from the symbol upload package's [PublishSymbols.targets](https://devdiv.visualstudio.com/DevDiv/_git/DotNet-SymUploader?path=%2Fsrc%2FMicrosoft.SymbolUploader.Build.Task%2FPackageFiles%2FPublishSymbols.targets&version=GBmaster)?

The `PublishSymbols` target in current PR and the linked `PublishSymbols` target don't do the same thing. I can replace the current call to PublishSymbols **task** to calling PublishSymbols target on the linked file if that is an issue.

dagood commented:
Yeah, that's the duplication I meant, up to you if you think this isn't a maintenance concern. I don't have particular reasons to think the task's API will change, I may have overreacted when I saw duplication on top of duplication.

JohnTortugo commented:
> The symbol packages do also need to be published to a build storage location so they can be reuploaded later at release time to extend the expiration time. (Not sure if that's what you were saying.)

Silly question: you mean the .symbols.nupkg? They get published together with the .nupkgs.

dagood commented:
Yep, thanks for confirming. (With https://github.com/dotnet/arcade/issues/2415 from the CR thread, though, we might not always have to do that.)

JohnTortugo commented:
I think I've addressed all your comments. Is there any other thing that I should address here?

I built this PR in these builds: 
- https://dnceng.visualstudio.com/internal/_build/results?buildId=141026
- https://dnceng.visualstudio.com/internal/_build/results?buildId=142168 
- https://dnceng.visualstudio.com/internal/_build/results?buildId=142152

I tested it's SDK in these builds:

- https://dnceng.visualstudio.com/internal/_build/results?buildId=142168
- https://dnceng.visualstudio.com/internal/_build/results?buildId=141026

JohnTortugo commented:
Tested the symbol publishing in this pipeline / releases:

- https://dnceng.visualstudio.com/internal/_releaseProgress?_a=release-pipeline-progress&releaseId=4155
- https://dnceng.visualstudio.com/internal/_releaseProgress?_a=release-pipeline-progress&releaseId=4110

JohnTortugo commented:
Anything else?
